scipy/linalg/_cythonized_array_utils.pyx

```diff
@@ -193,11 +193,13 @@ cdef inline (int, int) band_check_internal_c(const np_numeric_t[:, ::1]A) noexce
             break
 
     # upper triangular part
-    for r in range(n-1):
-        for c in range(m - 1, r + upper_band, -1):
+    for c in range(m - 1, 0, -1):
+        # Only bother if outside the existing band:
+        for r in range(min(c - upper_band, n - 1)):
```

Comment:
We cannot change the order of the `for` loops, because going through row by row is the fast lane since C-layout is row-major. That's the reason we start by lower left for the lower triangle and upper right for the upper triangle for quick exit. 

In other words, we scan the most obvious entries first however in the memory row elements are next to each other and column elements are `n` apart.

Reply:
understood. for a column major matrix, this code would have been good. restoring the order of for loops and doing the changes.

Reply:
Please let me know if it looks good now.

Reply:
Indeed that is the scan direction. I think you can `break` immediately after the previous iteration without spinning another time.Currently, you compute the `upper_band` after the `if`and wait for it to spin again (which is not a big deal obviously) but might affect the compiler black-magic.

Could you also add the test matrix that was failing before this PR under this test?

https://github.com/scipy/scipy/blob/b4f3037e04bd0d312cd671b7d1d5d9ffd8923472/scipy/linalg/tests/test_cythonized_array_utils.py#L28-L37

Reply:
You don't need to do in-place stuff like above. You can directly add the array without going over PEP8 line limit. 

` ` `python
    A = np.array([[1, 1, 0....]])
    assert bandwidth(A) == (2, 2)
` ` `

Reply:
@ilayn the test is running fine. Please let me know if there is any other scope of improvement.

Reply:
just realized that if I move the following to the end I need to increment r by 1 else it would be the same as the previous case of spinning an extra time. 

```
# If existing band falls outside matrix; we are done
        if r + upper_band > m - 1:
            break
```
 So it should be 
 
 ```
 # If existing band falls outside matrix; we are done
        if r + 1 + upper_band > m - 1:
            break
```

Reply:
Let me know if this looks good now. updated the PR.